# №1

In [26]:
from keras.datasets import imdb

(train_data, train_labels), (test_data, test_labels) = imdb.load_data(
    num_words=10000)

In [27]:
import numpy as np

def vectorize_sequences(sequences, dimension=10000):
    results = np.zeros((len(sequences), dimension))
    for i, sequence in enumerate(sequences):
        results[i, sequence] = 1.
    return results

x_train = vectorize_sequences(train_data)
x_test = vectorize_sequences(test_data)

In [29]:
y_train = np.asarray(train_labels).astype('float32')
y_test = np.asarray(test_labels).astype('float32')

In [30]:
from keras import optimizers
from keras import models
from keras import layers

model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))
model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss='binary_crossentropy',
              metrics=['accuracy'])

In [31]:
x_val = x_train[:10000]
partial_x_train = x_train[10000:]
y_val = y_train[:10000]
partial_y_train = y_train[10000:]

In [7]:
model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test) 
results

Epoch 1/4
25000/25000 [==============================] - 654s 26ms/step - loss: 0.4483 - acc: 0.8173
Epoch 2/4
25000/25000 [==============================] - 106s 4ms/step - loss: 0.2546 - acc: 0.9096
Epoch 3/4
25000/25000 [==============================] - 4s 167us/step - loss: 0.1968 - acc: 0.9296
Epoch 4/4
25000/25000 [==============================] - 871s 35ms/step


[0.2950650653219223, 0.88352]

## Home Tasks

The following experiments will help convince you that the architecture choices you’ve made are all fairly reasonable, although there’s still room for improvement:

### 1) You used two hidden layers. Try using one or three hidden layers, and see how doing so affects validation and test accuracy.

In [32]:
model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))
model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss='binary_crossentropy',
              metrics=['accuracy'])
model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test) 
results

Epoch 1/4
 1024/25000 [>.............................] - ETA: 12:30 - loss: 0.6891 - acc: 0.5625

D:\Programs\Anaconda3 5.2.0\lib\site-packages\keras\callbacks.py:122: UserWarning: Method on_batch_end() is slow compared to the batch update (0.117707). Check your callbacks.
  % delta_t_median)


25000/25000 [==============================] - 840s 34ms/step - loss: 0.4751 - acc: 0.8092
Epoch 2/4
25000/25000 [==============================] - 15s 587us/step - loss: 0.2629 - acc: 0.9075
Epoch 3/4
25000/25000 [==============================] - 3s 136us/step - loss: 0.1986 - acc: 0.9292
Epoch 4/4
25000/25000 [==============================] - 1206s 48ms/step


[0.3012056277227402, 0.88232]

In [34]:
model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(1, activation='sigmoid'))
model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss='binary_crossentropy',
              metrics=['accuracy'])
model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test) 
results

Epoch 1/4
25000/25000 [==============================] - 1427s 57ms/step - loss: 0.4433 - acc: 0.8277
Epoch 2/4
25000/25000 [==============================] - 31s 1ms/step - loss: 0.2714 - acc: 0.9089
Epoch 3/4
25000/25000 [==============================] - 4s 142us/step - loss: 0.2146 - acc: 0.9265
Epoch 4/4
25000/25000 [==============================] - 783s 31ms/step


[0.28360106026649473, 0.88732]

### Conclusion: 
    one hidden layer - [0.28360106026649473, 0.88732]
    two hidden layers - [0.2950650653219223, 0.88352] 
    three hidden layers - [0.3012056277227402, 0.88232]
    
    При уменьшении количества скрытых слоёв незначительно уменьшается test loss и повышается test accuracy

### 2) Try using layers with more hidden units or fewer hidden units: 32 units, 64 units, and so on.

In [35]:
model = models.Sequential()
model.add(layers.Dense(8, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(8, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))
model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss='binary_crossentropy',
              metrics=['accuracy'])
model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test) 
results

Epoch 1/4
25000/25000 [==============================] - 524s 21ms/step - loss: 0.5406 - acc: 0.8078
Epoch 2/4
25000/25000 [==============================] - 5s 212us/step - loss: 0.3445 - acc: 0.8937
Epoch 3/4
25000/25000 [==============================] - 3s 126us/step - loss: 0.2521 - acc: 0.9147
Epoch 4/4
25000/25000 [==============================] - 500s 20ms/step


[0.2787383371067047, 0.89052]

In [36]:
model = models.Sequential()
model.add(layers.Dense(128, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(128, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))
model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss='binary_crossentropy',
              metrics=['accuracy'])
model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test) 
results

Epoch 1/4
25000/25000 [==============================] - 322s 13ms/step - loss: 0.4253 - acc: 0.8052
Epoch 2/4
25000/25000 [==============================] - 7s 284us/step - loss: 0.2250 - acc: 0.9113
Epoch 3/4
25000/25000 [==============================] - 6s 255us/step - loss: 0.1616 - acc: 0.9365
Epoch 4/4
25000/25000 [==============================] - 404s 16ms/step


[0.35319236679553984, 0.87748]

### Conclusion:
    8 hidden units - [0.2787383371067047, 0.89052]
    16 hidden units - [0.2950650653219223, 0.88352]
    128 hidden units - [0.35319236679553984, 0.87748]
    
    При уменьшении количества скрытых единиц уменьшается test loss и незначительно увеличивается test accuracy.

### 3) Try using the mse loss function instead of binary_crossentropy.

In [37]:
model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))
model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss='mse',
              metrics=['accuracy'])
model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test) 
results

Epoch 1/4
25000/25000 [==============================] - 423s 17ms/step - loss: 0.1504 - acc: 0.8286
Epoch 2/4
25000/25000 [==============================] - 5s 193us/step - loss: 0.0797 - acc: 0.9075
Epoch 3/4
25000/25000 [==============================] - 3s 131us/step - loss: 0.0596 - acc: 0.9308
Epoch 4/4
25000/25000 [==============================] - 469s 19ms/step


[0.0853683164858818, 0.8852]

### Comclusion:
    binary_crossentropy - [0.2950650653219223, 0.88352]
    mse - [0.0853683164858818, 0.8852]
    При использовании mse значительно уменьшается test loss.

### 4) Try using the tanh activation (an activation that was popular in the early days of neural networks) instead of relu.

In [38]:
model = models.Sequential()
model.add(layers.Dense(16, activation='tanh', input_shape=(10000,)))
model.add(layers.Dense(16, activation='tanh'))
model.add(layers.Dense(1, activation='sigmoid'))
model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss='binary_crossentropy',
              metrics=['accuracy'])
model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test) 
results

Epoch 1/4
25000/25000 [==============================] - 1030s 41ms/step - loss: 0.4301 - acc: 0.8260
Epoch 2/4
25000/25000 [==============================] - 4s 157us/step - loss: 0.2344 - acc: 0.9132
Epoch 3/4
25000/25000 [==============================] - 4s 145us/step - loss: 0.1783 - acc: 0.9336
Epoch 4/4
25000/25000 [==============================] - 479s 19ms/step


[0.3241392808437347, 0.8772]

### Conclusion:
    relu activation - [0.2950650653219223, 0.88352]
    tanh activation - [0.3241392808437347, 0.8772]
    При использовании активации tanh test loss увеличивается.

# №2

In [10]:
from keras.datasets import reuters

(train_data, train_labels), (test_data, test_labels) = reuters.load_data(
    num_words=10000)

In [9]:
import numpy as np

def vectorize_sequences(sequences, dimension=10000):
    results = np.zeros((len(sequences), dimension))
    for i, sequence in enumerate(sequences):
        results[i, sequence] = 1.
    return results

x_train = vectorize_sequences(train_data)
x_test = vectorize_sequences(test_data)

def to_one_hot(labels, dimension=46):
    results = np.zeros((len(labels), dimension))
    for i, label in enumerate(labels):
        results[i, label] = 1.
    return results

one_hot_train_labels = to_one_hot(train_labels)
one_hot_test_labels = to_one_hot(test_labels)

from keras.utils.np_utils import to_categorical

one_hot_train_labels = to_categorical(train_labels)
one_hot_test_labels = to_categorical(test_labels)

x_val = x_train[:1000]
partial_x_train = x_train[1000:]

y_val = one_hot_train_labels[:1000]
partial_y_train = one_hot_train_labels[1000:]

In [11]:
from keras import models
from keras import layers

model = models.Sequential()
model.add(layers.Dense(64, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(46, activation='softmax'))
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
model.fit(partial_x_train,
          partial_y_train,
          epochs=9,
          batch_size=512,
          validation_data=(x_val, y_val))
results = model.evaluate(x_test, one_hot_test_labels)
results

Train on 7982 samples, validate on 1000 samples
Epoch 1/9
7982/7982 [==============================] - 10s 1ms/step - loss: 2.5241 - acc: 0.4977 - val_loss: 1.7178 - val_acc: 0.6120
Epoch 2/9
7982/7982 [==============================] - 2s 212us/step - loss: 1.4442 - acc: 0.6889 - val_loss: 1.3496 - val_acc: 0.7090
Epoch 3/9
7982/7982 [==============================] - 2s 211us/step - loss: 1.0992 - acc: 0.7643 - val_loss: 1.1752 - val_acc: 0.7430
Epoch 4/9
7982/7982 [==============================] - 2s 225us/step - loss: 0.8728 - acc: 0.8160 - val_loss: 1.0836 - val_acc: 0.7590
Epoch 5/9
7982/7982 [==============================] - 2s 209us/step - loss: 0.7060 - acc: 0.8488 - val_loss: 0.9866 - val_acc: 0.7810
Epoch 6/9
7982/7982 [==============================] - 2s 199us/step - loss: 0.5695 - acc: 0.8790 - val_loss: 0.9414 - val_acc: 0.8040
Epoch 7/9
7982/7982 [==============================] - 2s 207us/step - loss: 0.4621 - acc: 0.9038 - val_loss: 0.9078 - val_acc: 0.8040
Epoch 8/

[0.9823264784825666, 0.7853962600708857]

## Further experiments (Hometask)



### 1) Try using larger or smaller layers: 32 units, 128 units, and so on.

In [12]:
model = models.Sequential()
model.add(layers.Dense(32, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(32, activation='relu'))
model.add(layers.Dense(46, activation='softmax'))
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
model.fit(partial_x_train,
          partial_y_train,
          epochs=9,
          batch_size=512,
          validation_data=(x_val, y_val))
results = model.evaluate(x_test, one_hot_test_labels)
results

Train on 7982 samples, validate on 1000 samples
Epoch 1/9
7982/7982 [==============================] - 2s 235us/step - loss: 3.0975 - acc: 0.4546 - val_loss: 2.3165 - val_acc: 0.6090
Epoch 2/9
7982/7982 [==============================] - 1s 178us/step - loss: 1.8966 - acc: 0.6625 - val_loss: 1.6682 - val_acc: 0.6590
Epoch 3/9
7982/7982 [==============================] - 1s 163us/step - loss: 1.4035 - acc: 0.7175 - val_loss: 1.4031 - val_acc: 0.6950
Epoch 4/9
7982/7982 [==============================] - 1s 165us/step - loss: 1.1576 - acc: 0.7502 - val_loss: 1.2633 - val_acc: 0.7280
Epoch 5/9
7982/7982 [==============================] - 1s 168us/step - loss: 0.9929 - acc: 0.7890 - val_loss: 1.1801 - val_acc: 0.7520
Epoch 6/9
7982/7982 [==============================] - 1s 167us/step - loss: 0.8635 - acc: 0.8166 - val_loss: 1.1190 - val_acc: 0.7720
Epoch 7/9
7982/7982 [==============================] - 1s 165us/step - loss: 0.7562 - acc: 0.8356 - val_loss: 1.0802 - val_acc: 0.7700
Epoch 8

[1.0722838923640263, 0.7644701692427468]

In [13]:
model = models.Sequential()
model.add(layers.Dense(128, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(128, activation='relu'))
model.add(layers.Dense(46, activation='softmax'))
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
model.fit(partial_x_train,
          partial_y_train,
          epochs=9,
          batch_size=512,
          validation_data=(x_val, y_val))
results = model.evaluate(x_test, one_hot_test_labels)
results

Train on 7982 samples, validate on 1000 samples
Epoch 1/9
7982/7982 [==============================] - 3s 346us/step - loss: 2.2531 - acc: 0.5534 - val_loss: 1.4524 - val_acc: 0.6630
Epoch 2/9
7982/7982 [==============================] - 2s 284us/step - loss: 1.1717 - acc: 0.7428 - val_loss: 1.1415 - val_acc: 0.7490
Epoch 3/9
7982/7982 [==============================] - 2s 279us/step - loss: 0.8202 - acc: 0.8260 - val_loss: 1.0116 - val_acc: 0.7760
Epoch 4/9
7982/7982 [==============================] - 2s 277us/step - loss: 0.5955 - acc: 0.8776 - val_loss: 0.9961 - val_acc: 0.7720
Epoch 5/9
7982/7982 [==============================] - 2s 276us/step - loss: 0.4333 - acc: 0.9114 - val_loss: 0.9036 - val_acc: 0.7960
Epoch 6/9
7982/7982 [==============================] - 2s 277us/step - loss: 0.3378 - acc: 0.9263 - val_loss: 0.8720 - val_acc: 0.8190
Epoch 7/9
7982/7982 [==============================] - 2s 277us/step - loss: 0.2604 - acc: 0.9411 - val_loss: 0.8659 - val_acc: 0.8270
Epoch 8

[1.1203325656727818, 0.7684772929652716]

### Conclusion:
    32 hidden units - [1.0722838923640263, 0.7644701692427468]
    64 hidden units - [0.9823264784825666, 0.7853962600708857]
    128 hidden units - [1.1203325656727818, 0.7684772929652716]
    
    Слои с 64 скрытыми единицами - самый низкий показатель test loss и самая высокая test accuracy.

### 2) You used two hidden layers. Now try using a single hidden layer, or three hidden layers.

In [14]:
model = models.Sequential()
model.add(layers.Dense(64, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(46, activation='softmax'))
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
model.fit(partial_x_train,
          partial_y_train,
          epochs=9,
          batch_size=512,
          validation_data=(x_val, y_val))
results = model.evaluate(x_test, one_hot_test_labels)
results

Train on 7982 samples, validate on 1000 samples
Epoch 1/9
7982/7982 [==============================] - 2s 269us/step - loss: 2.4502 - acc: 0.5531 - val_loss: 1.7298 - val_acc: 0.6450
Epoch 2/9
7982/7982 [==============================] - 2s 204us/step - loss: 1.4111 - acc: 0.7179 - val_loss: 1.3156 - val_acc: 0.7180
Epoch 3/9
7982/7982 [==============================] - 2s 209us/step - loss: 1.0441 - acc: 0.7874 - val_loss: 1.1265 - val_acc: 0.7540
Epoch 4/9
7982/7982 [==============================] - 2s 218us/step - loss: 0.8211 - acc: 0.8316 - val_loss: 1.0136 - val_acc: 0.7890
Epoch 5/9
7982/7982 [==============================] - 2s 206us/step - loss: 0.6597 - acc: 0.8703 - val_loss: 0.9498 - val_acc: 0.7920
Epoch 6/9
7982/7982 [==============================] - 2s 205us/step - loss: 0.5390 - acc: 0.8919 - val_loss: 0.8916 - val_acc: 0.8140
Epoch 7/9
7982/7982 [==============================] - 2s 204us/step - loss: 0.4415 - acc: 0.9126 - val_loss: 0.8568 - val_acc: 0.8190
Epoch 8

[0.88876136370461, 0.7978628673196795]

In [15]:
model = models.Sequential()
model.add(layers.Dense(64, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(46, activation='softmax'))
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
model.fit(partial_x_train,
          partial_y_train,
          epochs=9,
          batch_size=512,
          validation_data=(x_val, y_val))
results = model.evaluate(x_test, one_hot_test_labels)
results

Train on 7982 samples, validate on 1000 samples
Epoch 1/9
7982/7982 [==============================] - 2s 266us/step - loss: 2.5622 - acc: 0.5053 - val_loss: 1.6806 - val_acc: 0.6450
Epoch 2/9
7982/7982 [==============================] - 2s 205us/step - loss: 1.4219 - acc: 0.6972 - val_loss: 1.3588 - val_acc: 0.6900
Epoch 3/9
7982/7982 [==============================] - 2s 205us/step - loss: 1.1071 - acc: 0.7496 - val_loss: 1.2192 - val_acc: 0.7220
Epoch 4/9
7982/7982 [==============================] - 2s 206us/step - loss: 0.8902 - acc: 0.8014 - val_loss: 1.1329 - val_acc: 0.7350
Epoch 5/9
7982/7982 [==============================] - 2s 205us/step - loss: 0.7058 - acc: 0.8415 - val_loss: 1.0288 - val_acc: 0.7880
Epoch 6/9
7982/7982 [==============================] - 2s 206us/step - loss: 0.5670 - acc: 0.8756 - val_loss: 0.9874 - val_acc: 0.8020
Epoch 7/9
7982/7982 [==============================] - 2s 210us/step - loss: 0.4422 - acc: 0.9059 - val_loss: 1.0496 - val_acc: 0.7870
Epoch 8

[1.1389648417543325, 0.7653606411928804]

### Conclusion:
    one hidden layer - [0.88876136370461, 0.7978628673196795]
    two hidden layers - [0.9823264784825666, 0.7853962600708857]
    three hidden layers - [0.9823264784825666, 0.7853962600708857]
    
    Чем меньше слоёв, тем ниже test loss и выше test accuracy.